drivers/flash/spi_nand.c

```diff
@@ -0,0 +1,1034 @@
+{
+	int ret;
+	uint8_t secur_reg = 0;
+
+	acquire_device(dev);
+
+	ret = spi_nand_get_feature(dev, SPI_NAND_FEA_ADDR_CONF_B0, &secur_reg);
+	if (ret != 0) {
+		LOG_ERR("get feature failed: %d", ret);
+		goto out;
+	}
+
+	secur_reg |= SPINAND_SECURE_BIT_CONT;
+	ret = spi_nand_set_feature(dev, SPI_NAND_FEA_ADDR_CONF_B0, secur_reg);
+	if (ret != 0) {
+		LOG_ERR("set feature failed: %d", ret);
+		goto out;
+	}
+
+	ret = spi_nand_get_feature(dev, SPI_NAND_FEA_ADDR_CONF_B0, &secur_reg);
+	if (ret != 0) {
+		LOG_ERR("get feature failed: %d", ret);
+		goto out;
+	}
+
+	if (!(secur_reg & SPINAND_SECURE_BIT_CONT)) {
+		LOG_ERR("Enable continuous read failed: %d\n", secur_reg);
+	}
+
+out:	release_device(dev);
+	return ret;
+}
+
+static int spi_nand_conti_read_disable(const struct device *dev)
+{
+	int ret;
+	uint8_t secur_reg = 0;
+
+	acquire_device(dev);
+
+	ret = spi_nand_get_feature(dev, SPI_NAND_FEA_ADDR_CONF_B0, &secur_reg);
+	if (ret != 0) {
+		LOG_ERR("get feature failed: %d", ret);
+		goto out;
+	}
+
+	secur_reg &= ~SPINAND_SECURE_BIT_CONT;
+	ret = spi_nand_set_feature(dev, SPI_NAND_FEA_ADDR_CONF_B0, secur_reg);
+	if (ret != 0) {
+		LOG_ERR("set feature failed: %d", ret);
+		goto out;
+	}
+
+	ret = spi_nand_get_feature(dev, SPI_NAND_FEA_ADDR_CONF_B0, &secur_reg);
+	if (ret != 0) {
+		LOG_ERR("get feature failed: %d", ret);
+		goto out;
+	}
+
+	if (secur_reg & SPINAND_SECURE_BIT_CONT) {
+		LOG_ERR("Disable continuous read failed: %d\n", secur_reg);
+	}
+
+out:	release_device(dev);
+	return ret;
+}
+
+static int spi_nand_conti_read_exit(const struct device *dev)
+{
+	int ret;	
+	ret = spi_nand_cmd_write(dev, SPI_NAND_CMD_RESET);
+
+	return ret;
+}
+
+static int spi_nand_read_cont(const struct device *dev, off_t addr, void *dest,
+			size_t size)
+{
+	struct spi_nand_data *data = dev->data;
+	int ret = 0;
+
+	acquire_device(dev);
+
+	ret = spi_nand_access(dev, SPI_NAND_CMD_PAGE_READ,
+		NAND_ACCESS_ADDRESSED | NAND_ACCESS_24BIT_ADDR,
+		addr >> data->page_shift, NULL, 0);
+	if (ret != 0) {
+		LOG_ERR("page read failed: %d", ret);
+		goto out;
+	}
+
+	ret = spi_nand_wait_until_ready(dev);
+
+	ret = spi_nand_access(dev, SPI_NAND_CMD_READ_CACHE,
+		NAND_ACCESS_ADDRESSED | NAND_ACCESS_24BIT_ADDR,
+		((addr >> data->page_shift) & SPI_NAND_PAGE_MASK), dest, size);
+	if (ret != 0) {
+		LOG_ERR("read from cache failed: %d", ret);
+		goto out;
+	}
+
+	ret = spi_nand_conti_read_exit(dev);
+
+out:	release_device(dev);
+
+	return ret;
+}
+
+static int spi_nand_read_normal(const struct device *dev, 
+				off_t addr, void *dest, size_t size)
+{
+	struct spi_nand_data *data = dev->data;
+	int ret = 0;
+	uint32_t offset = 0;
+	uint32_t chunk = 0;
+	uint32_t read_bytes = 0;
+
+	acquire_device(dev);
+
+	while (size > 0) {
+		/* Read on _page_size_bytes boundaries (Default 2048 bytes a page) */
+		offset = addr % data->page_size;
+		chunk = (offset + size < data->page_size) ? 
+			size : (data->page_size - offset);
+		read_bytes = chunk;
```

Comment:
Why keep re-calculating this in loop? You can figure out the start, the and and the increment before entering the loop.